File: fastapi-graphql/app/movies.py
import requests, os
from dotenv import dotenv_values
from graphene import ObjectType, List, String, Schema, Field, Mutation, Int, Float, List, Boolean
from starlette.graphql import GraphQLApp
from schemas import MovieType, GenreType, GenreInputType, CompanyType, CompanyInputType
from .elasticsearch import es

if os.environ.get("TWITTER_API_TOKEN"):
    token = os.environ.get("TWITTER_API_TOKEN")
else:
    config = dotenv_values(".env")
    token = config["TWITTER_API_TOKEN"]

twitter_auth_header = {
    "authorization": f"Bearer {token}"
}

class MovieQuery(ObjectType):
    movie_list= None
    get_movie = Field(List(MovieType), id=Int(), title=String(), genre_name=String(), genre_id=Int(), company_id=Int(), company_name=String())
    async def resolve_get_movie(self, info, id=None, title=None, genre_name=None, genre_id=None, company_id=None, company_name=None):
        movie_list = []

        query_args = []
        if id:
            query_args.append({"match": {"id": id}})
        if title:
            query_args.append({"match": {"title": title}})
        if genre_name:
            query_args.append({"match": {"genres.name": genre_name}})
        if genre_id:
            query_args.append({"match": {"genres.id": genre_id}})
        if company_id:
            query_args.append({"match": {"production_companies.id": company_id}})
        if company_name:
            query_args.append({"match": {"production_companies.name": company_name}})

        if id or title or genre_name or genre_id or company_id or company_name:
            res = es.search(
                index="movies",
                body={
                    "size": 1000,
                    "query": {
                        "bool": {
                            "must": query_args
                        }
                    }
                }
            )
            for hit in res['hits']['hits']:
                movie_list.append(hit['_source'])
        else:
            res = es.search(
                index="movies",
                body={
                    "size": 1000,
                    "query": {
                        "match_all": {}
                    }
                }
            )
            for hit in res['hits']['hits']:
                movie_list.append(hit['_source'])

        return movie_list

class CheckMovie(Mutation):
    does_exist = Boolean()

    class Arguments:
        id = Int(required=True)

    async def mutate(self, info, id):
        does_exist = False

        res = es.search(
            index="movies", 
            body={
                "query": {
                    "match": {
                        "id": id
                    }
                }
            }
        )

        if res['hits']['total']['value'] == 0:
            does_exist = False
        if res['hits']['total']['value'] == 1:
            does_exist = True

        return CheckMovie(does_exist)

class AddMovie(Mutation):
    movie = Field(MovieType)

    class Arguments:
        id = Int(required=True)
        title = String(required=True)
        imdb_id = String(required=True)
        budget = Float(required=True)
        revenue = Float(required=True)
        vote_average = Float(required=True)
        genres = List(GenreInputType)
        vote_count = Int()
        runtime = Int()
        popularity = Float()
        production_companies = List(CompanyInputType)
        twitter_id = String()

    async def mutate(self, info, id, title, imdb_id, budget, revenue, vote_average, genres, vote_count=None, runtime=None, popularity=None, production_companies=None, twitter_id=None):
        twitter_follower = None

        if twitter_id:
            response = requests.get(f'https://api.twitter.com/1.1/users/show.json?screen_name={twitter_id}', headers=twitter_auth_header)
            account = response.json()
            if "followers_count" in account:
                twitter_follower = account["followers_count"]
        
        new_movie = {
            'id': id,
            'title': title,
            'imdb_id': imdb_id,
            'budget': budget,
            'revenue': revenue,
            'vote_average': vote_average,
            'vote_count': vote_count,
            'runtime': runtime,
            'popularity': popularity,
            'genres': genres,
            'production_companies': production_companies,
            'twitter_id': twitter_id,
            'twitter_follower': twitter_follower
        }
        res = es.index(
            index="movies", 
            body=new_movie
        )
        return AddMovie(new_movie)

class MovieMutation(ObjectType):
    add_movie = AddMovie.Field()
    check_movie = CheckMovie.Field()

File: react/src/api/graphql/movies.js
import { GraphQLClient, gql } from 'graphql-request'

const gqlClient = new GraphQLClient(`${process.env.REACT_APP_GRAPHQL_BASE_URL}/movies`)

const stringify = (objectArray) => {
    let string = "["
    objectArray.map((obj, index) => {
        string += "{"
        let keys = Object.keys(obj)
        keys.map((key, index) => {
            string += key + ": "
            Number.isInteger(obj[key]) ? string += `${obj[key]}` : string += `"${obj[key]}"`
            keys.length-1 !== index && (string += ",")
        })
        objectArray.length - 1 !== index ? string += "}," : string += "}"
    })
    string += "]"
    return string
}

export const addMovie = async (payload) => {
    try {
        const mutation = gql`
            mutation {
                addMovie(id: ${payload.id}, imdbId: "${payload.imdbId}", title: "${payload.title}", budget: ${payload.budget}, revenue: ${payload.revenue}, runtime: ${payload.runtime}, voteAverage: ${payload.vote}, voteCount: ${payload.voteCount}, popularity: ${payload.popularity}, genres: ${stringify(payload.genres)}, productionCompanies: ${stringify(payload.productionCompanies)}, ${payload.twitter_id ? (`, twitterId: "${payload.twitter_id}"`) : ""}) {
                    movie {
                        id
                        title
                        imdbId
                        budget
                        revenue
                        popularity
                        genres {
                            id
                            name
                        }
                        voteCount
                        voteAverage
                        runtime
                        productionCompanies {
                            id
                            name
                        }
                        ${payload.imgPath ? "twitterId" : ""}
                    }
                }
            }
        `;

        const response = await gqlClient.request(mutation)
        return response
    } catch(e) {
        console.error(e)
        return null
    }
}

export const checkMovie = async (id) => {
    try {
        const mutation = gql`
            mutation {
                checkMovie(id: ${id}) {
                    doesExist
                }
            }
        `;

        const response = await gqlClient.request(mutation)
        return response
    } catch(e) {
        console.error(e)
        return null
    }
}

export const getMoviesByKey = async (payload) => {
    try {
        let queryArgs = []
        if(payload.name) queryArgs.push(`name: "${payload.name}"`)
        if(payload.knownForDep) queryArgs.push(`knownForDep: "${payload.knownForDep}"`)
        if(payload.id) queryArgs.push(`id: ${payload.id}`)
        if(payload.imdbId) queryArgs.push(`imdbId: "${payload.imdbId}"`)

        const query = gql`
            query {
                getPeople(${queryArgs.join()}) {
                    id,
                    imdbId,
                    title,
                    budget,
                    revenue,
                    voteAverage,
                    voteCount,
                    runtime,
                    popularMovies,
                    imgPath,
                    twitterId,
                    twitterFollower
                }
            }
        `;

        const response = await gqlClient.request(query)
        return response
    } catch(e) {
        console.error(e)
        return null
    }
}